src/re_graph/core.cljc

```diff
@@ -34,10 +34,41 @@
                                 :payload {:query query
                                           :variables variables}}]}))))
 
-(defn mutate [& args]
+#?(:clj
+   (defn ^:private sync-wrapper
+     "Wraps the given function to allow the GraphQL result to be returned
+      synchronously. Will return a GraphQL error response if the request times
+      out. Will throw if the call returns an exception."
+     [f timeout & args]
+     (let [p        (promise)
+           callback (fn [result] (deliver p result))
+           args'    (conj (vec args) callback)]
+       (apply f args')
+
+       ;; explicit timeout to avoid unreliable aborts from underlying implementations
+       (let [result (deref p timeout :timeout)]
+         (if (= :timeout result)
+           {:errors [{:status 500, :message "GraphQL request timed out", :args args}]}
+           result)))))
+
+(defn mutate
+  "Execute a GraphQL mutation. The arguments are:
+
+  [instance-name query-string variables callback]
+
+  If the optional `instance-name` is not provided, the default instance is
+  used. The callback function will receive the result of the mutation as its
+  sole argument."
+  [& args]
   (let [callback-fn (last args)]
     (re-frame/dispatch (into [::mutate] (conj (vec (butlast args)) [::internals/callback callback-fn])))))
 
+#?(:clj
+   (def ^{:doc "Executes a mutation synchronously. Takes the same arguments as
+                `mutate` but without the callback."}
+     mutate-sync
+     (partial sync-wrapper mutate 3000)))    ;; 3s timeout
```

Comment:
I think the timeout should either be an additional arg to the `-sync` functions and/or else be specified in the options passed to `init` (although it might be tricky to read it out when needed). 3s seems very arbitrary.

Reply:
I've added an optional timeout at the end of argument list for `mutate-sync` and `query-sync`.